### PR TITLE
Move Personal Power and Energon to System values

### PR DESCRIPTION
**Pull request by JFranklin3000**
##### In this PR
- Moved personal Power and Energon to system variables
- Added Sorcerous Power as a system variable. 
- Migration for Personal Power and Energon added
-Added Energon to AltMode Tab

##### Testing
- Create Transformer and validate energon migration on the actor
- Create Power Ranger and validate Personal Power Migration
- Validate casting cost is still working for powers

##### Next Change
-Remove Energon and Personal Power creation from actor.mjs

### CODE REVIEW

File: template.json [hunk 1]
```diff
@@ -91,6 +95,17 @@
             "total": 0
           }
         },
+        "power": {
```
Comment (phildominguez) [inline template.json:98]: If these are going to be grouped, I would go with plural `powers`

File: templates/actor/parts/items/power/container.hbs [hunk 2]
```diff
@@ -11,9 +30,9 @@
 
   {{#*inline "item-label" item}}
     <strong>
-      {{item.name}}:
+      {{item.name}}
       {{#ifEquals @root.actor.type "npc"}}
-        {{#if system.usesPer}}
+        :{{#if system.usesPer}}
```
Comment (phildominguez) [inline templates/actor/parts/items/power/container.hbs:35]: How come?

File: module/documents/actor.mjs [hunk 3]
```diff
@@ -94,6 +94,7 @@ export class Essence20Actor extends Actor {
     // Make separate methods for each Actor type (character, npc, etc.) to keep
     // things organized.
     this._prepareNpcData();
+    this._prepareSorcerousPower();
     if (["giJoe", "pony", "powerRanger", "transformer"].includes(this.type)) {
```
Comment (phildominguez) [inline module/documents/actor.mjs:98]:
```suggestion
    this._prepareSorcerousPower();

    if (["giJoe", "pony", "powerRanger", "transformer"].includes(this.type)) {
```

File: module/migration.mjs [hunk 4]
```diff
@@ -215,6 +215,17 @@ export const migrateActorData = function(actor) {
 
     if (itemToDelete.type == "threatPower") {
       await itemToDelete.delete();
+    } else if (itemToDelete.type == "classFeature") {
+      console.log (itemToDelete);
```
Comment (phildominguez) [inline module/migration.mjs:219]:
```suggestion
```

File: module/migration.mjs [hunk 5]
```diff
@@ -215,6 +215,17 @@ export const migrateActorData = function(actor) {
...
+      if (itemToDelete.name == "Personal Power") {
+        updateData[`system.power.personal.max`] = itemToDelete.system.uses.max;
+        updateData[`system.power.personal.value`] = itemToDelete.system.uses.value;
+        await itemToDelete.delete();
+      } else if (itemToDelete.name == "Energon") {
+        updateData[`system.energon.value`] = itemToDelete.system.uses.value;
+        updateData[`system.energon.max`] = itemToDelete.system.uses.max;
+        itemToDelete.delete();
```
Comment (phildominguez) [inline module/migration.mjs:227]: Noticing a mix of deletes with and without `await`s. Just checking they're only being used when needed.

File: module/sheet-handlers/perk-handler.mjs [hunk 6]
```diff
@@ -24,14 +24,10 @@ export class PerkHandler {
...
-          type: 'classFeature',
-          data: {},
-        },
-        { parent: this._actor},
-      );
+      await this._actor.update ({
+        "system.power.sorcerous.levelTaken": this._actor.system.level,
+      });
+
     }
```
Comment (phildominguez) [inline module/sheet-handlers/perk-handler.mjs:31]:
```suggestion
      });
    }
```

File: module/sheet-handlers/power-handler.mjs [hunk 7]
```diff
@@ -21,16 +21,10 @@ export class PowerHandler {
...
     for (let actorItem of this._actor.items) {
-      if (actorItem.type =='classFeature') {
-        if (power.system.type == "grid" && actorItem.name == 'Personal Power') {
-          classFeatureId = actorItem._id;
-        } else if (power.system.type == "sorcerous" && actorItem.name == 'Sorcerous Power') {
-          classFeatureId = actorItem._id;
-        }
-      } else if (actorItem.type == 'power' && actorItem.system.originalId == powerUuid) {
+
+      if (actorItem.type == 'power' && actorItem.system.originalId == powerUuid) {
```
Comment (phildominguez) [inline module/sheet-handlers/power-handler.mjs:27]:
```suggestion
      if (actorItem.type == 'power' && actorItem.system.originalId == powerUuid) {
```

File: module/sheet-handlers/power-handler.mjs [hunk 8]
```diff
@@ -54,20 +47,27 @@ export class PowerHandler {
...
+      powerType = "threat";
     }
 
-    if (power.system.hasVariableCost && classFeature) {
+    // for (const actorEffect of this._actor.effects) {
+    //   const parts = actorEffect.origin.split(".");
+    //   if (parts[3] == power._id) {
+    //     actorEffect.update({disabled: false});
+    //   }
+    // }
```
Comment (phildominguez) [inline module/sheet-handlers/power-handler.mjs:64]: Is this being fixed later?

File: module/sheet-handlers/power-handler.mjs [hunk 9]
```diff
@@ -80,13 +80,14 @@ export class PowerHandler {
...
               label: game.i18n.localize('E20.AcceptButton'),
-              callback: html => this.powerCountUpdate(rememberValues(html), power, classFeature),
+              callback: html => this.powerCountUpdate(rememberValues(html), power, powerType),
             },
           },
         },
       ).render(true);
-    } else if (classFeature && classFeature.system.uses.value >= power.system.powerCost){
-      classFeature.update({ ["system.uses.value"]: Math.max(0, classFeature.system.uses.value - power.system.powerCost) });
+    } else if (powerType != "threat" && this._actor.system.power[powerType].value >= power.system.powerCost){
```
Comment (phildominguez) [inline module/sheet-handlers/power-handler.mjs:88]:
```suggestion
    } else if (powerType != "threat" && this._actor.system.power[powerType].value >= power.system.powerCost) {
```

File: module/sheet-handlers/power-handler.mjs [hunk 10]
```diff
@@ -100,14 +101,15 @@ export class PowerHandler {
   * @param {Power} power The power
   * @param {ClassFeature} classFeature  The classFeature that is tied to the power
```
Comment (phildominguez) [inline module/sheet-handlers/power-handler.mjs:102]: Update @param

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: templates/actor/parts/items/power/container.hbs (hunk 2 above)
Comment (JFranklin3000) [reply templates/actor/parts/items/power/container.hbs:35]: So it only shows up on the NPC sheet. The character sheets in the book don't have that descriptor.

File: module/migration.mjs (hunk 5 above)
Comment (JFranklin3000) [reply module/migration.mjs]: That one was missing I added it.

File: module/sheet-handlers/power-handler.mjs (hunk 8 above)
Comment (JFranklin3000) [reply module/sheet-handlers/power-handler.mjs:63]: It is actually working but I need to add a way to turn off effects when they expire or the combat ends, before I consider it done.

File: module/sheet-handlers/power-handler.mjs (hunk 10 above)
Comment (JFranklin3000) [reply module/sheet-handlers/power-handler.mjs:101]: updated.

File: template.json (hunk 1 above)
Comment (JFranklin3000) [reply template.json:98]: changed.

File: templates/actor/parts/items/power/container.hbs (hunk 2 above)
Comment (phildominguez) [reply templates/actor/parts/items/power/container.hbs:35]: Makes sense, what about putting the `:` by itself on the line above? It just looks kinda funky right on an if-statement

File: templates/actor/parts/items/power/container.hbs [hunk 11]
```diff
@@ -1,3 +1,22 @@
...
+  {{/if}}
+    {{#if system.powers.sorcerous.levelTaken}}
+  <label>Sorcerous Power</label>
+    <input class="centered-input no-lock" data-field="system.powers.sorcerous.value" type="number"
+      name="system.powers.sorcerous.value" value="{{system.powers.sorcerous.value}}" />
+    /
+    <input class="centered-input no-lock" data-field="system.powers.sorcerous.max" type="number" name="system.powers.sorcerous.max"
+      value="{{system.powers.sorcerous.max}}" />
+  {{/if}}
+</div>
```
Comment (phildominguez) [inline templates/actor/parts/items/power/container.hbs:18]: This file should just be for the item container. This stats-container should be in its own template, then added in pc-body.hbs. You can see a blank container on the NPC sheet, and making that change will fix this. Similar for altMode.

**Review by phildominguez — CHANGES_REQUESTED** [review]

**Review by phildominguez — APPROVED** [review]

**Pull request merged**